### PR TITLE
New @sigstore/mock package

### PR DESCRIPTION
**Pull request by bdehamer**
#### Summary
Package providing mocking support for the core APIs in Fulcio, Rekor and the Timestamp Authority.

Unlike the main `sigstore` project where we've made an intentional effort to limit our reliance on third-party dependencies, I've included all sorts of libraries here to help with the ugly work of generating certificates/timestamps, canonicalizing JSON, parsing OIDC tokens, etc. This package should only ever be used as a `devDependency` so I'm not concerned with dependency bloat.

### CODE REVIEW

File: packages/mock/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+* Mocked version of the Sigstore Fulcio `POST /api/v2/signingCert` API which
+  returns a verifiable certificate signed by an ephemeral certificate authority.
+* Mocked version of the Sigstore Rekor `POST /api/v1/log/entries` API which
+  returns a log entry with a verifiable signed-entry timestamp (SET).
+* Mocked version of the Sigstore Timestamp Authority `POST /api/v1/timestamp`
+  API which returns a verifiable signed timestamp.
+
+## To Do
+
+* Mocked TUF repository which returns the key material necessary to
```
Comment (feelepxyz) [inline packages/mock/README.md:17]: This is going to be so cool!

File: packages/mock/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+interface TSAOptions {
+  baseURL?: string;
+  strict?: boolean;
+}
+
+export async function mockFulcio(options: FulcioOptions = {}) {
+  const url = options.baseURL || DEFAULT_FULCIO_URL;
+  const strict = options.strict ?? true;
```
Comment (feelepxyz) [inline packages/mock/src/index.ts:43]: Nice :)

File: packages/mock/src/fulcio/ca.test.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    describe('when a custom extension is provided', () => {
+      it('issues a cert with the extension', async () => {
+        const ca = await initializeCA();
+
+        // Issue a certificate with the key above
+        const signingCert = await ca.issueCertificate({
+          publicKey: keyBytes,
+          subjectAltName: 'test',
+          extensions: [extension],
```
Comment (feelepxyz) [inline packages/mock/src/fulcio/ca.test.ts:150]: Being able to set custom extensions is very nice!

File: packages/mock/src/fulcio/ctlog.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    issuerID: ArrayBuffer
+  ): Promise<ArrayBuffer> {
+    const version = 0;
+    const timestamp = new Date();
+    const logID = this.logID;
+
+    // A bit of a hack to get the TBS bytes
+    const tbs = pkijs.Certificate.fromBER(certificate).encodeTBS().toBER();
+
+    // Calculate the SCT signature
```
Comment (feelepxyz) [inline packages/mock/src/fulcio/ctlog.ts:93]: Could we link to something where this is implemented in fulcio/ctlog for reference? Thinking down the line there might be some drift between implementations so having a way to find where things are implemented could be helpful.

File: packages/mock/src/fulcio/ca.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    // convert key to CryptoKey
+    const key = await this.crypto.subtle.importKey(
+      'spki',
+      opts.publicKey,
+      KEY_ALGORITHM_ECDSA_P256,
+      true,
+      ['sign', 'verify']
+    );
+
+    const tbs = {
```
Comment (feelepxyz) [inline packages/mock/src/fulcio/ca.ts:103]: Anything useful we could link to in fulcio for the actual implementation for ref?

File: packages/mock/src/rekor/tlog.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      .createHash('sha256')
+      .update(this.publicKey)
+      .digest()
+      .toString('hex');
+    const body = canonicalize(proposedEntry);
+
+    // Random number for logIndex
+    const logIndex = Math.floor(Math.random() * 9000000) + 1000000;
+
+    // Calculate SET
```
Comment (feelepxyz) [inline packages/mock/src/rekor/tlog.ts:69]: Do we have anything we can link to in rekor for ref?

File: packages/mock/src/timestamp/tsa.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    this.rootCert = options.rootCertificate;
+    this.keyPair = options.keyPair;
+    this.crypto = new pkijs.CryptoEngine({ crypto: new Crypto() });
+  }
+
+  public get rootCertificate(): Buffer {
+    return Buffer.from(this.rootCert.toSchema().toBER(false));
+  }
+
+  public async timestamp(req: TimestampRequest): Promise<Buffer> {
```
Comment (feelepxyz) [inline packages/mock/src/timestamp/tsa.ts:75]: Is this basically implementing the bare bones of the tsa rfc? Worth linking to any of it?

**Review by feelepxyz — DISMISSED** [review]
This is really great! Would be so cool if we could get to the point where we can drive some kind of conformance test suite with this.

Left some nice to have comments if there are useful references we can link to for reference.

**Comment by bdehamer** [comment]
@feelepxyz I liked your suggestion to leave some links in the code to relevant implementation specs. Updated with links to the RFCs or matching code in Fulcio/Rekor

**Review by feelepxyz — APPROVED** [review]

**Pull request merged**